source/common/router/router.h

```diff
@@ -165,6 +165,13 @@ class Filter : Logger::Loggable<Logger::Id::router>,
   }
   const Router::MetadataMatchCriteria* metadataMatchCriteria() override {
     if (route_entry_) {
+      // Have we been called before? If so, there's no need to recompute because
+      // by the time this method is called for the first time, route_entry_ should
+      // not change anymore.
+      if (metadata_match_ != nullptr) {
```

Comment:
Can you add a test for this as well? Thanks.

Reply:
@htuch done — plus two bonus tests:

* using > 1 key/value pair in dynamicMetadata()
* mixing a key that overrides the route entry with one that doesn't

thanks for the review!